File: findAnagrams/findAnagrams.py
from collections import Counter
from typing import List

def findAnagrams(s: str, p: str) -> List[int]:
    result = []
    window = Counter()
    p_freq = Counter(p)

    left, right = 0, 0

    while right < len(s):
        window[s[right]] += 1
        right += 1

        if right - left > len(p):
            window[s[left]] -= 1
            if window[s[left]] == 0:
                del window[s[left]]
            left += 1

        if right - left == len(p):
            if window == p_freq:
                result.append(left)

    return result

s = "cbaebabacd"
p = "abc"
print(findAnagrams(s, p))

File: findAnagrams/findAnagrams.cpp
#include <iostream>
#include <vector>
#include <string>
#include <array>

using namespace std;

vector<int> findAnagrams(string s, string p) {
    vector<int> result;
    array<int, 26> window = {0};
    array<int, 26> pFreq = {0};

    for (char c : p) {
        pFreq[c - 'a']++;
    }

    int left = 0, right = 0;

    while (right < s.size()) {
        window[s[right] - 'a']++;
        right++;

        if (right - left > p.size()) {
            window[s[left] - 'a']--;
            left++;
        }

        if (right - left == p.size()) {
            bool isAnagram = true;
            for (int i = 0; i < 26; ++i) {
                if (window[i] != pFreq[i]) {
                    isAnagram = false;
                    break;
                }
            }
            if (isAnagram) {
                result.push_back(left);
            }
        }
    }

    return result;
}

int main() {
    string s = "cbaebabacd";
    string p = "abc";

    vector<int> result = findAnagrams(s, p);

    cout << "Indices of anagrams: ";
    for (int idx : result) {
        cout << idx << " ";
    }
    cout << endl;

    return 0;
}
